cmd/gateway-manta.go

```diff
@@ -0,0 +1,481 @@
+	"github.com/joyent/triton-go/authentication"
+	"github.com/joyent/triton-go/storage"
+	"github.com/minio/cli"
+	"github.com/minio/minio/pkg/hash"
+)
+
+// stor is a namespace within manta where you store any documents that are deemed as private
+// and require access credentials to read them. Within the stor namespace, you can create any
+// number of directories and objects.
+const (
+	mantaDefaultRootStore = "/stor"
+	mantaBackend          = "manta"
+	defaultMantaURL       = "https://us-east.manta.joyent.com"
+)
+
+func init() {
+	const mantaGatewayTemplate = `NAME:
+  {{.HelpName}} - {{.Usage}}
+USAGE:
+  {{.HelpName}} {{if .VisibleFlags}}[FLAGS]{{end}} [ENDPOINT]
+{{if .VisibleFlags}}
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}{{end}}
+ENDPOINT:
+  Manta server endpoint. Default ENDPOINT is https://us-east.manta.joyent.com
+ENVIRONMENT VARIABLES:
+  ACCESS:
+     MINIO_ACCESS_KEY: The Manta account name.
+     MINIO_SECRET_KEY: A KeyID associated with the Manta account.
+  BROWSER:
+     MINIO_BROWSER: To disable web browser access, set this value to "off".
+EXAMPLES:
+  1. Start minio gateway server for Manta Object Storage backend.
+      $ export MINIO_ACCESS_KEY=manta_account_name
+      $ export MINIO_SECRET_KEY=manta_key_id
+      $ {{.HelpName}}
+  2. Start minio gateway server for Manta Object Storage backend on custom endpoint.
+      $ export MINIO_ACCESS_KEY=manta_account_name
+      $ export MINIO_SECRET_KEY=manta_key_id
+      $ {{.HelpName}} https://us-west.manta.joyent.com
+
+`
+
+	MustRegisterGatewayCommand(cli.Command{
+		Name:               mantaBackend,
+		Usage:              "Manta Object Storage.",
+		Action:             mantaGatewayMain,
+		CustomHelpTemplate: mantaGatewayTemplate,
+		Flags:              append(serverFlags, globalFlags...),
+		HideHelpCommand:    true,
+	})
+}
+
+func mantaGatewayMain(ctx *cli.Context) {
+	host := ctx.Args().First()
+	startGateway(ctx, &MantaGateway{host})
+}
+
+// MantaGateway implements Gateway.
+type MantaGateway struct {
+	host string
+}
+
+// Name implements Gateway interface.
+func (g *MantaGateway) Name() string {
+	return mantaBackend
+}
+
+// NewGatewayLayer returns manta gateway layer, implements GatewayLayer interface to
+// talk to manta remote backend.
+func (g *MantaGateway) NewGatewayLayer() (GatewayLayer, error) {
+	log.Println(colorYellow("\n               *** Warning: Not Ready for Production ***"))
+	return newMantaGateway(g.host)
+}
+
+// tritonObjects - Implements Object layer for Triton Manta storage
+type tritonObjects struct {
+	gatewayUnsupported
+	client *storage.StorageClient
+}
+
+func newMantaGateway(host string) (GatewayLayer, error) {
+	var err error
+	var signer authentication.Signer
+	var endpoint = defaultMantaURL
+
+	if host != "" {
+		endpoint, _, err = parseGatewayEndpoint(host)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	creds := serverConfig.GetCredential()
+	keyMaterial := os.Getenv("MINIO_KEY_MATERIAL")
```

Comment:
This should be named as MANTA_KEY_MATERIAL, since this has nothing to do with MINIO.